memory_allocator.c

```diff
@@ -0,0 +1,258 @@
+#include <common.h>
+#include <stdint.h>
+#include <stddef.h>
+#include "queue.h"
+#include "libkern.h"
+#include "include/memory_allocator.h"
+
+/*
+  TODO:
+  - use the mp_next field of malloc_pool
+*/
+
+#define MAGIC 0xC0DECAFE
+#define ALIGNMENT 8
+
+TAILQ_HEAD(mb_list, mem_block);
+TAILQ_HEAD(mp_arena, mem_arena);
+
+static inline int32_t abs(int32_t x)
+{
+  if (x < 0)
+    return -x;
+  return x;
+}
+
+typedef struct mem_block {
+  uint32_t mb_magic; /* if overwritten report a memory corruption error */
+  int32_t mb_size; /* size < 0 => free, size > 0 => alloc'd */
+  TAILQ_ENTRY(mem_block) mb_list;
```

Comment:
Hmm... is there any good reason to keep the list of used blocks ? Perhaps just a list of free blocks would do, and we could save some space by:
```
union {
  TAILQ_ENTRY(mem_block) mb_list;
  uint64_t mb_data[0];
};
```
What do you think ?